### PR TITLE
feat(pool): Add entry point routing to pool

### PR DESCRIPTION
**Pull request by dancoombs**
## Proposed Changes

  - Removes some of the extra generics on PoolOperation from a previous PR
  - Routes to the correct mempool in the pool server
  - Track multiple entry points in chain tracking

### CODE REVIEW

File: crates/pool/src/chain.rs [hunk 1]
```diff
@@ -416,68 +434,113 @@ impl<P: Provider> Chain<P> {
...
+                        "Log with unknown entry point address: {:?}. Ignoring.",
+                        log.address
+                    );
+                    continue;
+                }
             }
         }
 
-        mined_ops
+        Ok((mined_ops, entity_balance_updates))
```
Comment (dancoombs) [inline crates/pool/src/chain.rs:457]: TODO: we should collect these by entry point address

File: crates/pool/src/chain.rs [hunk 2]
```diff
@@ -1028,7 +1094,7 @@ mod tests {
             address: ENTRY_POINT_ADDRESS,
             topics: vec![
                 H256::from(utils::keccak256(
-                    UserOperationEventFilter::abi_signature().as_bytes(),
+                    entry_point_v0_6::UserOperationEventFilter::abi_signature().as_bytes(),
```
Comment (dancoombs) [inline crates/pool/src/chain.rs:1097]: TODO: add tests for v0.7

File: crates/pool/src/server/local.rs [hunk 3]
```diff
@@ -564,8 +546,10 @@ where
                         // Async methods
                         // Responses are sent in the spawned task
                         ServerRequestKind::AddOp { entry_point, op, origin } => {
-                            let fut = |mempool: Arc<M>, response: oneshot::Sender<Result<ServerResponse, PoolError>>| async move {
-                                let resp = match mempool.add_operation(origin, op.into()).await {
+                            let fut = |mempool: Arc<Box<dyn Mempool>>, response: oneshot::Sender<Result<ServerResponse, PoolError>>| async move {
+                                // TODO check entrypoint version and OP
```
Comment (dancoombs) [inline crates/pool/src/server/local.rs:550]: TODO: check op type before sending to the pool, else things will crash.

File: crates/pool/src/chain.rs (hunk 1 above)
Comment (dancoombs) [reply crates/pool/src/chain.rs:457]: Actually not going to do this, the list should be small so an iteration/filter shouldn't be a problem

File: crates/pool/src/chain.rs [hunk 4]
```diff
@@ -416,68 +434,113 @@ impl<P: Provider> Chain<P> {
...
-            if let Ok(event) = contract::parse_log::<DepositedFilter>(log.clone()) {
-                let info = BalanceUpdate {
-                    entrypoint,
-                    address: event.account,
-                    amount: event.total_deposit,
-                    is_addition: true,
-                };
-
-                balance_updates.push(info);
+    fn load_v0_6(log: Log, mined_ops: &mut Vec<MinedOp>, balance_updates: &mut Vec<BalanceUpdate>) {
```
Comment (0xfourzerofour) [inline crates/pool/src/chain.rs:460]: This is a nice compared to the previous

**Review by 0xfourzerofour — APPROVED** [review]
LGTM

File: crates/pool/src/task.rs [hunk 5]
```diff
@@ -155,19 +190,23 @@ impl PoolTask {
...
         provider: Arc<P>,
-    ) -> anyhow::Result<
-        UoPool<
-            v0_6::UserOperation,
-            impl Prechecker<UO = v0_6::UserOperation>,
-            impl Simulator<UO = v0_6::UserOperation>,
-            impl EntryPoint,
-        >,
-    > {
+    ) -> anyhow::Result<Arc<Box<dyn Mempool>>> {
```
Comment (dphilipson) [inline crates/pool/src/task.rs:209]: I've sort of forgotten Rust a bit, but can this just be `Arc<dyn Mempool>` without the `Box`?

**Review by dphilipson — APPROVED** [review]
Looks great!

**Pull request merged**